store/sqlstore/channel_store.go

```diff
@@ -3659,73 +3670,83 @@ func (s SqlChannelStore) UpdateSidebarCategories(userId, teamId string, categori
 	if err != nil {
 		return nil, model.NewAppError("SqlChannelStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.open_transaction.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
-
 	defer finalizeTransaction(transaction)
-	for _, categoryToUpdate := range categories {
-		originalCategory, appErr := s.GetSidebarCategory(categoryToUpdate.Id)
+
+	updatedCategories := []*model.SidebarCategoryWithChannels{}
+	for _, category := range categories {
+		originalCategory, appErr := s.GetSidebarCategory(category.Id)
 		if appErr != nil {
 			return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, appErr.Error(), http.StatusInternalServerError)
 		}
 
-		category := &originalCategory.SidebarCategory
-
-		// we can update display name only on custom categories
-		if category.DisplayName != categoryToUpdate.DisplayName {
-			if category.Type != model.SidebarCategoryCustom {
-				return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.invalid_update_fields", map[string]interface{}{"field": "DisplayName"}, "", http.StatusBadRequest)
-			}
-			category.DisplayName = categoryToUpdate.DisplayName
+		// Copy category to avoid modifying an argument
+		updatedCategory := &model.SidebarCategoryWithChannels{
+			SidebarCategory: category.SidebarCategory,
 		}
 
-		category.Sorting = categoryToUpdate.Sorting
+		// Prevent any changes to read-only fields of SidebarCategories
+		updatedCategory.UserId = originalCategory.UserId
+		updatedCategory.TeamId = originalCategory.TeamId
+		updatedCategory.SortOrder = originalCategory.SortOrder
+		updatedCategory.Type = originalCategory.Type
 
-		if _, err = transaction.UpdateColumns(func(col *gorp.ColumnMap) bool {
-			return col.ColumnName == "DisplayName" || col.ColumnName == "Sorting"
-		}, category); err != nil {
-			return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+		if updatedCategory.Type != model.SidebarCategoryCustom {
+			updatedCategory.DisplayName = originalCategory.DisplayName
 		}
 
-		// clean previous SidebarChannels
-		sql, args, _ := s.getQueryBuilder().Delete("SidebarChannels").Where(sq.Eq{"ChannelId": categoryToUpdate.Channels}).ToSql()
+		if category.Type != model.SidebarCategoryDirectMessages {
+			updatedCategory.Channels = make([]string, len(category.Channels))
+			copy(updatedCategory.Channels, category.Channels)
+		}
 
-		if _, err = transaction.Exec(sql, args...); err != nil {
+		if _, err = transaction.UpdateColumns(func(col *gorp.ColumnMap) bool {
```

Comment:
You said in the yesterday meeting the `UpdateColumns` is error prone and have some drawbacks, why no use an sql approach, like using Squirrel to build a query that clearly express the intention. "Update SidebarCategories SET DisplayName = X, Sorting = Y WHERE Id = Z"?

Reply:
Just in case anyone comes across this later, the followup PR #14843 addresses this